src/compiler/crystal/compiler.cr

```diff
@@ -300,6 +303,18 @@ module Crystal
 
         %(#{CL} #{object_name} "/Fe#{output_filename}" #{program.lib_flags} #{link_flags})
       else
+        if thin_lto
+          lto_cache_dir = "#{output_dir}/lto.cache"
+          Dir.mkdir_p(lto_cache_dir)
+          {% if flag?(:darwin) %}
+            cc = ENV["CC"]? || "clang-4.0 -flto=thin -Wl,-mllvm,-threads=#{n_threads},-cache_path_lto,#{lto_cache_dir},#{@release ? "-mllvm,-O2" : "-mllvm,-O0"}"
```

Comment:
What about `clang-5.0` or `clang-6.0` (nightly)?

Maybe just expecting `clang`, as it was initialy, and document that a `clang` binary must exist (e.g. as a link to `clang-4.0` or `clang-5.0`)? Or maybe accept an `ENV["CLANG"]`.